### DISCUSSION TITLE
[performance+memory]  Beating git in `index-pack` (as used for clones and fetches) ✅ 🚀

### DISCUSSION THREAD

**Discussion by Byron** [GitoxideLabs/gitoxide#46] [Show and tell]
git index-pack is streaming a pack and creates an index from it. The difficulty arises from having to decompress every entry in the pack stream, which can be composed of many small objects. These are placed in some sort of index to accelerate the next stage that is all about resolving the deltas in order to produce a SHA1. Per pack entry, the SHA1, pack offset and CRC32 are written into the index file to complete the operation.

The indexing phase in inherently single-threaded with little potential for improvements, whereas the resolving phase is fully multithreaded and entirely _lock free_. The first phase could be improved by writing the pack file in parallel - right now it happens after reading it (_the pack file is used later for lookup to not hold everything in memory_). However, IO doesn't appear to be the bottleneck at all.

Compared to `gitoxide`, git is considerably faster when creating the index, averaging 54MB/s of reading uncompressed bytes. `gitoxide` clocks in at about ~~45MB/s~~ 50MB/s, and slows down considerably during the end. Part of that slowdown might be attributed [to this issue with resetting miniz_oxide's decompressor](https://github.com/Frommi/miniz_oxide/issues/89).

Luckily `gitoxide` is way faster when resolving deltas, which already gives it a good first place in the race, with some room for more if it manages to get as fast as git when decompressing and indexing objects.

The picture below shows the fastest git run I could produce, probably with everything being properly cached:

<img width="1302" alt="Screenshot 2020-08-04 at 12 04 36" src="https://user-images.githubusercontent.com/63622/89252979-e4f04980-d64d-11ea-9106-c3e992dbf907.png">

Without cache, it seems to look different:

<img width="1302" alt="Screenshot 2020-08-04 at 12 04 36" src="https://user-images.githubusercontent.com/63622/89253041-0f420700-d64e-11ea-9e8e-1b7dd49f94a8.png">

The fastest `gitoxide` runs, which are pretty comparable in the amount of work done, as they also write out the pack and the index. The only difference is that they use the packfile directly instead of reading it from stdin, it's streamed nonetheless though, and merely an oversight.

<img width="1304" alt="Screenshot 2020-08-04 at 12 28 45" src="https://user-images.githubusercontent.com/63622/89253099-3698d400-d64e-11ea-87a8-db278a3dd3db.png">

Memory consumption of git hovers consistently around 650MB (for the kernel pack), and is ~~lower~~ higher than the ~~1.2GB~~ ~~750MB~~ 580MB that `gitoxide` uses. However, `gitoxide` can temporarily use more memory as it keeps intermediate decompressed objects per thread, whose maximum sizes depend on the amount of children and the base size. So I have seen this go up to 850MB for small fractions of time because of that.

**Comment by Byron** [comment]
#### memory mode: in-memory, resolve-bases, resolve-deltas, resolve-deltas-and-bases

More like a test I wanted to see if it makes any difference to keep the decompressed data in memory to speed up downstream operations.

And it looks like this is actually reducing performance at least while the pack is also streamed to disk at the same time. The virtual memory system probably caches it entirely.

<img width="1298" alt="Screenshot 2020-08-04 at 12 42 52" src="https://user-images.githubusercontent.com/63622/89253806-3699d380-d650-11ea-8ea9-19c0a8eb2ea4.png">

When not streaming the pack to disk, in-memory operation appears to be yielding a mild speedup. But when allowing to write a temporary file, the speedup is entirely gone. Thus it seems that keeping decompressed bytes really doesn't do any good.

<img width="1295" alt="Screenshot 2020-08-04 at 12 51 15" src="https://user-images.githubusercontent.com/63622/89254338-8036ee00-d651-11ea-8f08-2219ddcaa1ea.png">

<details>

```
➜  grit-rs git:(main) ✗ time ./target/release/gixp -t 4 index-from-pack  --verbose -i verify -m in-memory out < tests/fixtures/repos/rust.git/objects/pack/pack-af2bf75bff7478224bd4242e5106241ab6026232.pack
 04:47:38 indexing done 1266143 objects in 14.51s (87286.266 objects/s)
 04:47:48 Resolving done 1266143 objects in 9.76s (129665.88 objects/s)
 04:47:49 writing index file done 35453056 bytes in 0.13s (262973120 bytes/s)
 04:47:49  create index file done 1266143 objects in 24.97s (50702.586 objects/s)
./target/release/gixp -t 4 index-from-pack --verbose -i verify -m in-memory    46.92s user 4.46s system 205% cpu 24.992 total
➜  grit-rs git:(main) ✗ time ./target/release/gixp -t 4 index-from-pack  --verbose -i verify -m in-memory < tests/fixtures/repos/rust.git/objects/pack/pack-af2bf75bff7478224bd4242e5106241ab6026232.pack
 04:48:29 indexing done 1266143 objects in 13.19s (95986.98 objects/s)
 04:48:38 Resolving done 1266143 objects in 9.49s (133432.3 objects/s)
 04:48:39 writing index file done 35453056 bytes in 0.08s (433581380 bytes/s)
 04:48:39  create index file done 1266143 objects in 23.34s (54253.984 objects/s)
./target/release/gixp -t 4 index-from-pack --verbose -i verify -m in-memory <  47.60s user 2.85s system 216% cpu 23.354 total
➜  grit-rs git:(main) ✗ time ./target/release/gixp -t 4 index-from-pack  --verbose -i verify < tests/fixtures/repos/rust.git/objects/pack/pack-af2bf75bff7478224bd4242e5106241ab6026232.pack             04:50:53 indexing done 1266143 objects in 13.64s (92843.51 objects/s)
 04:51:02 Resolving done 1266143 objects in 9.32s (135791.89 objects/s)
 04:51:03 writing index file done 35453056 bytes in 0.08s (429502050 bytes/s)
 04:51:03  create index file done 1266143 objects in 23.64s (53559.31 objects/s)
./target/release/gixp -t 4 index-from-pack --verbose -i verify <   46.50s user 3.01s system 208% cpu 23.693 total
➜  grit-rs git:(main) ✗
```

**Comment by Byron** [comment]
For the actual performance tests on a 96 core machine, have a look at [this comment](https://github.com/Byron/gitoxide/issues/1#issuecomment-672566602). tldr;: the time is dominated by creating an index by streaming the pack, and pack resolution is then done in about 10 seconds or 14.6GB/s (of decoded objects).

**Comment by Byron** [comment]
The ARM git provided with MacOS Big Sur changes everything:

With 3 threads (default)
```
➜  gitoxide git:(main) time git index-pack --stdin -v < ./tests/fixtures/repos/linux.git/objects/pack/pack-3ee05b0f4e4c2cb59757c95c68e2d13c0a491289.pack
Receiving objects: 100% (7600359/7600359), 1.26 GiB | 82.23 MiB/s, done.
Resolving deltas: 100% (6396700/6396700), done.
pack    3ee05b0f4e4c2cb59757c95c68e2d13c0a491289
git index-pack --stdin -v <   75.78s user 5.78s system 199% cpu 40.845 total
```

Git is at least twice as fast when reading/streaming the pack. In our case this is limited by the deflate performance of millions of small streams, and there are still some improvements that we can make use of.

With 8 threads (as available cores)
```
➜  gitoxide git:(main) time git -c pack.threads=8 index-pack --stdin -v < ./tests/fixtures/repos/linux.git/objects/pack/pack-3ee05b0f4e4c2cb59757c95c68e2d13c0a491289.pack
Receiving objects: 100% (7600359/7600359), 1.26 GiB | 49.73 MiB/s, done.
Resolving deltas: 100% (6396700/6396700), done.
pack    3ee05b0f4e4c2cb59757c95c68e2d13c0a491289
git -c pack.threads=8 index-pack --stdin -v <   129.59s user 29.72s system 174% cpu 1:31.07 total
```

Clearly contention reduces speed. This effect is not visible at all when verifying pack entries, making me think the amount of work `git` does is entirely different (by now).

`gitoxide` looses whenever many small objects need to be decompressed, which is clearly visible in the later stages of the decoding stage. It also uses more CPU time that is disproportional to the speed gain.

```
➜  gitoxide git:(main) time ./target/release/gixp --verbose pack-index-from-data -p tests/fixtures/repos/linux.git/objects/pack/pack-3ee05b0f4e4c2cb59757c95c68e2d13c0a491289.pack out
 16:19:30 read pack done 1.4GB in 29.48s (46.2MB/s)
 16:19:30  indexing done 7.6M objects in 29.48s (257.8k objects/s)
 16:19:30 decompressing done 2.7GB in 29.48s (92.2MB/s)
 16:19:57     Resolving done 7.6M objects in 26.78s (283.9k objects/s)
 16:19:57      Decoding done 95.6GB in 26.78s (3.6GB/s)
index: 3fe49647a452e0f7ee6a857fb65ee243f6e38bb3==============================================================================================================>-----------------------------------------]
pack: 3ee05b0f4e4c2cb59757c95c68e2d13c0a491289=======================================================================>---------------------------------------------------------------------------------]
 16:19:59 writing index file done 212.8MB in 0.42s (506.2MB/s)
 16:19:59  create index file done 7.6M objects in 57.86s (131.4k objects/s)
./target/release/gixp --verbose pack-index-from-data -p  out  226.00s user 3.44s system 396% cpu 57.911 total
```

**Reply by Byron** [reply]
Fortunately, even though it seems special thanks to the `(Apple Git-128)` suffix, a git built from source is just as fast (here with 4 threads).

```
➜  gitoxide git:(main) git --version
git version 2.24.3 (Apple Git-128)
➜  gitoxide git:(main) ✗
➜  gitoxide git:(main) ✗ time  ~/dev/github.com/git/git/git -c pack.threads=4 index-pack  -v --show-resolving-progress tests/fixtures/repos/linux.git/objects/pack/pack-3ee05b0f4e4c2cb59757c95c68e2d13c0a491289.pack
Indexing objects: 100% (7600359/7600359), done.
Resolving deltas: 100% (6396700/6396700), done.
3ee05b0f4e4c2cb59757c95c68e2d13c0a491289
~/dev/github.com/git/git/git -c pack.threads=4 index-pack -v    77.93s user 7.05s system 238% cpu 35.610 total
➜  gitoxide git:(main) ✗ ~/dev/github.com/git/git/git --version                                                                                                                                         git version 2.31.0
```

**Comment by Byron** [comment]
#### About the Arrival of ZLibNG and ARM64 performance

As there have been plenty of updates sprinkled in multiple discussions, let me sum up the current state here and provide some hopefully reproducible runs on an M1 MacBook Air using a 1.3GB Linux kernel pack on commit 2b5d89154fd8d3f1934af007a33fa1d20360ec15 .

Recently a major improvement was done which brings `flate2` support directly to `gitoxide`, allowing various backends to be used. The fastest is `zlib-ng`, which delivers a 33%  performance improvement, with the speed of the indexing phase being **doubled**.

Before with `miniz-oxide`…

```
mkdir -p out
git checkout fb52a24ab40d575649857c28c478c551c560756f
make release-lean
time ./target/release/gixp -t 8 --verbose pack-index-from-data -p tests/fixtures/repos/linux.git/objects/pack/pack-3ee05b0f4e4c2cb59757c95c68e2d13c0a491289.pack out
```

```
 05:29:24 read pack done 1.4GB in 31.74s (42.9MB/s)
 05:29:24  indexing done 7.6M objects in 31.74s (239.4k objects/s)
 05:29:24 decompressing done 2.7GB in 31.74s (85.6MB/s)
 05:29:54     Resolving done 7.6M objects in 29.37s (258.8k objects/s)
 05:29:54      Decoding done 95.6GB in 29.37s (3.3GB/s)
index: 3fe49647a452e0f7ee6a857fb65ee243f6e38bb3
pack: 3ee05b0f4e4c2cb59757c95c68e2d13c0a491289
 05:29:55 writing index file done 212.8MB in 0.45s (470.1MB/s)
 05:29:55  create index file done 7.6M objects in 62.72s (121.2k objects/s)
./target/release/gixp -t 8 --verbose pack-index-from-data -p  out  247.03s user 3.50s system 398% cpu 1:02.81 total

```

…and after with `zlib-ng` as `flate2` backend. _(Please note that the latter also supports the `miniz-oxide` backend and it may soon be an option to turn that on using feature toggles)_.

```
git checkout 2b5d89154fd8d3f1934af007a33fa1d20360ec15
make release-lean
time ./target/release/gixp -t 8 --verbose pack-index-from-data -p tests/fixtures/repos/linux.git/objects/pack/pack-3ee05b0f4e4c2cb59757c95c68e2d13c0a491289.pack out
```

```
 13:14:56 read pack done 1.4GB in 15.34s (88.7MB/s)
 13:14:56  indexing done 7.6M objects in 15.34s (495.5k objects/s)
 13:14:56 decompressing done 2.7GB in 15.34s (177.1MB/s)
 13:15:23     Resolving done 7.6M objects in 26.33s (288.7k objects/s)
 13:15:23      Decoding done 95.6GB in 26.33s (3.6GB/s)
index: 3fe49647a452e0f7ee6a857fb65ee243f6e38bb3
pack: 3ee05b0f4e4c2cb59757c95c68e2d13c0a491289
 13:15:24 writing index file done 212.8MB in 0.47s (449.5MB/s)
 13:15:24  create index file done 7.6M objects in 43.29s (175.6k objects/s)
./target/release/gixp -t 8 --verbose pack-index-from-data -p  out  208.94s user 3.24s system 489% cpu 43.321 total
```

The indexing phase is now twice as fast and roughly en-par with `git`, even though pack resolution on ARM is still faster with 4 performance cores at least.

```
time  git -c pack.threads=4 index-pack  -v --show-resolving-progress tests/fixtures/repos/linux.git/objects/pack/pack-3ee05b0f4e4c2cb59757c95c68e2d13c0a491289.pack
Indexing objects: 100% (7600359/7600359), done.
Resolving deltas: 100% (6396700/6396700), done.
3ee05b0f4e4c2cb59757c95c68e2d13c0a491289
git -c pack.threads=4 index-pack -v --show-resolving-progress   75.66s user 5.45s system 239% cpu 33.819 total
```

However, as opposed to `gitoxide`, `git` can't scale as well with more cores on x86, even though it does seem to scale well on ARM. Here is the same with 8 cores (involving the high-efficiency cores)…

```
time  git -c pack.threads=8 index-pack  -v --show-resolving-progress tests/fixtures/repos/linux.git/objects/pack/pack-3ee05b0f4e4c2cb59757c95c68e2d13c0a491289.pack
Indexing objects: 100% (7600359/7600359), done.
Resolving deltas: 100% (6396700/6396700), done.
3ee05b0f4e4c2cb59757c95c68e2d13c0a491289
git -c pack.threads=8 index-pack -v --show-resolving-progress   91.87s user 19.36s system 357% cpu 31.137 total
```

…and here is the default with 3 threads …

```
 time  git  index-pack  -v --show-resolving-progress tests/fixtures/repos/linux.git/objects/pack/pack-3ee05b0f4e4c2cb59757c95c68e2d13c0a491289.pack
Indexing objects: 100% (7600359/7600359), done.
Resolving deltas: 100% (6396700/6396700), done.
3ee05b0f4e4c2cb59757c95c68e2d13c0a491289
git index-pack -v --show-resolving-progress   75.10s user 4.44s system 208% cpu 38.179 total
```

It shows that the ARM version of git does something truly outstanding whose 3 threads are easily outperforming `gitoxide` using 8 threads (4 performance cores, 4 efficiency cores), even though as far as I know it does stop scaling nicely with more cores at least on x86 machines. That's where `gitoxide` truly shines and I wonder what's the root cause of this (_note that with nightly as of 2021-03-21 there is no improvement in runtime_).

**Reply by Byron** [reply]
I will keep some 'notes to self' here to find out how git decodes pack to try understanding why it's faster than gitoxide (even though that only seems to be the case on ARM64).

* [it seems to open the pack once per thread](https://github.com/git/git/blob/98164e9585e02e31dcf1377a553efe076c15f8c6/builtin/index-pack.c#L190:L190)
* [threads take objects off a central object queue and there is a lock that might slow things down under contention](https://github.com/git/git/blob/98164e9585e02e31dcf1377a553efe076c15f8c6/builtin/index-pack.c#L1026:L1026)

**Reply by Byron** [reply]
After all… a profiling session revealed that `gitoxide` spends most of its time in `sha1::compress::compress`, so a faster sha1 might be a solution here. Apparently the one for ARM is just considerably slower than on intel at least when using the `sha-1` crate. `git` doesn't suffer from that issue.

<img width="903" alt="Screenshot 2021-03-22 at 09 44 16" src="https://user-images.githubusercontent.com/63622/111930248-6cae9980-8af3-11eb-962d-ab50c79709e6.png">

**Reply by edward-shen** [reply]
How are you profiling? I was trying to get some flamegraphs running but couldn't get meaningful data on that. Is this a mac only solution?

**Reply by Byron** [reply]
When profiling `git` we find this picture (focussing on the resolving step):

<img width="643" alt="Screenshot 2021-03-22 at 10 05 29" src="https://user-images.githubusercontent.com/63622/111931427-24dd4180-8af6-11eb-8b64-a59c5b578dad.png">

It uses something faster in `libcorecrypto.dylib` apparently, so that might be an avenue at some point worth noting in `performance.md`.

**Reply by Byron** [reply]
It's good to see that git spends way more time with other things so I am quite confident that with a sufficiently fast SHA1 implementation on ARM64, the tides can be turned there as well. On X86 that wasn't an issue last time I tested.

@edward-shen Yes, that's a Mac tool called 'Instruments'.

**Reply by Byron** [reply]
There is the "asm" feature that I didn't enable in the "sha-1" crate in "git-features". When doing so, I see the following unfortunately:


```
cargo build --release
   Compiling sha1-asm v0.4.4
The following warnings were emitted during compilation:

warning: src/aarch64.S:63:2: error: ADR/ADRP relocations must be GOT relative
warning:  adrp x1, .K0
warning:  ^
warning: src/aarch64.S:63:2: error: unknown AArch64 fixup kind!
warning:  adrp x1, .K0
warning:  ^
warning: src/aarch64.S:64:2: error: unknown AArch64 fixup kind!
warning:  ldr q4, [x1, #:lo12:.K0]
warning:  ^
warning: src/aarch64.S:101:2: error: ADR/ADRP relocations must be GOT relative
warning:  adrp x1, .K1
warning:  ^
warning: src/aarch64.S:101:2: error: unknown AArch64 fixup kind!
warning:  adrp x1, .K1
warning:  ^
warning: src/aarch64.S:102:2: error: unknown AArch64 fixup kind!
warning:  ldr q4, [x1, #:lo12:.K1]
warning:  ^
warning: src/aarch64.S:140:2: error: ADR/ADRP relocations must be GOT relative
warning:  adrp x1, .K2
warning:  ^
warning: src/aarch64.S:140:2: error: unknown AArch64 fixup kind!
warning:  adrp x1, .K2
warning:  ^
warning: src/aarch64.S:141:2: error: unknown AArch64 fixup kind!
warning:  ldr q4, [x1, #:lo12:.K2]
warning:  ^
warning: src/aarch64.S:179:2: error: ADR/ADRP relocations must be GOT relative
warning:  adrp x1, .K3
warning:  ^
warning: src/aarch64.S:179:2: error: unknown AArch64 fixup kind!
warning:  adrp x1, .K3
warning:  ^
warning: src/aarch64.S:180:2: error: unknown AArch64 fixup kind!
warning:  ldr q4, [x1, #:lo12:.K3]
warning:  ^

error: failed to run custom build command for `sha1-asm v0.4.4`

Caused by:
  process didn't exit successfully: `/Users/byron/dev/github.com/Byron/gitoxide/target/release/build/sha1-asm-3b930b340039d957/build-script-build` (exit code: 1)
  --- stdout
  TARGET = Some("aarch64-apple-darwin")
  OPT_LEVEL = Some("3")
  HOST = Some("aarch64-apple-darwin")
  CC_aarch64-apple-darwin = None
  CC_aarch64_apple_darwin = None
  HOST_CC = None
  CC = None
  CFLAGS_aarch64-apple-darwin = None
  CFLAGS_aarch64_apple_darwin = None
  HOST_CFLAGS = None
  CFLAGS = None
  CRATE_CC_NO_DEFAULTS = None
  DEBUG = Some("false")
  CARGO_CFG_TARGET_FEATURE = None
  running: "cc" "-O3" "-ffunction-sections" "-fdata-sections" "-fPIC" "-arch" "arm64" "-Wall" "-Wextra" "-march=armv8-a+crypto" "-c" "-o" "/Users/byron/dev/github.com/Byron/gitoxide/target/release/build/sha1-asm-37b139d871ad77c8/out/src/aarch64.o" "-c" "src/aarch64.S"
  cargo:warning=src/aarch64.S:63:2: error: ADR/ADRP relocations must be GOT relative
  cargo:warning= adrp x1, .K0
  cargo:warning= ^
  cargo:warning=src/aarch64.S:63:2: error: unknown AArch64 fixup kind!
  cargo:warning= adrp x1, .K0
  cargo:warning= ^
  cargo:warning=src/aarch64.S:64:2: error: unknown AArch64 fixup kind!
  cargo:warning= ldr q4, [x1, #:lo12:.K0]
  cargo:warning= ^
  cargo:warning=src/aarch64.S:101:2: error: ADR/ADRP relocations must be GOT relative
  cargo:warning= adrp x1, .K1
  cargo:warning= ^
  cargo:warning=src/aarch64.S:101:2: error: unknown AArch64 fixup kind!
  cargo:warning= adrp x1, .K1
  cargo:warning= ^
  cargo:warning=src/aarch64.S:102:2: error: unknown AArch64 fixup kind!
  cargo:warning= ldr q4, [x1, #:lo12:.K1]
  cargo:warning= ^
  cargo:warning=src/aarch64.S:140:2: error: ADR/ADRP relocations must be GOT relative
  cargo:warning= adrp x1, .K2
  cargo:warning= ^
  cargo:warning=src/aarch64.S:140:2: error: unknown AArch64 fixup kind!
  cargo:warning= adrp x1, .K2
  cargo:warning= ^
  cargo:warning=src/aarch64.S:141:2: error: unknown AArch64 fixup kind!
  cargo:warning= ldr q4, [x1, #:lo12:.K2]
  cargo:warning= ^
  cargo:warning=src/aarch64.S:179:2: error: ADR/ADRP relocations must be GOT relative
  cargo:warning= adrp x1, .K3
  cargo:warning= ^
  cargo:warning=src/aarch64.S:179:2: error: unknown AArch64 fixup kind!
  cargo:warning= adrp x1, .K3
  cargo:warning= ^
  cargo:warning=src/aarch64.S:180:2: error: unknown AArch64 fixup kind!
  cargo:warning= ldr q4, [x1, #:lo12:.K3]
  cargo:warning= ^
  exit code: 1

  --- stderr


  error occurred: Command "cc" "-O3" "-ffunction-sections" "-fdata-sections" "-fPIC" "-arch" "arm64" "-Wall" "-Wextra" "-march=armv8-a+crypto" "-c" "-o" "/Users/byron/dev/github.com/Byron/gitoxide/target/release/build/sha1-asm-37b139d871ad77c8/out/src/aarch64.o" "-c" "src/aarch64.S" with args "cc" did not execute successfully (status code exit code: 1).

```

Maybe that's something to sort out with the maintainers of the  `hashes` crate.

**Comment by Byron** [comment]
Here are the fastest runs on an M1 with the latest version.

```
➜  gitoxide git:(main) /usr/bin/time -lp  ./target/release/gixp -t 8 --verbose pack-index-from-data -p  tests/fixtures/repos/linux.git/objects/pack/pack-3ee05b0f4e4c2cb59757c95c68e2d13c0a491289.pack out
 15:21:04 read pack done 1.4GB in 15.33s (88.8MB/s)
 15:21:04  indexing done 7.6M objects in 15.33s (495.8k objects/s)
 15:21:04 decompressing done 2.7GB in 15.33s (177.2MB/s)
 15:21:30     Resolving done 7.6M objects in 26.61s (285.6k objects/s)
 15:21:30      Decoding done 95.6GB in 26.61s (3.6GB/s)
 15:21:32 writing index file done 212.8MB in 0.48s (445.6MB/s)
index: 3fe49647a452e0f7ee6a857fb65ee243f6e38bb3==============================================================================================================>-----------------------------------------]
pack: 3ee05b0f4e4c2cb59757c95c68e2d13c0a491289
 15:21:32  create index file done 7.6M objects in 44.06s (172.5k objects/s)
real        44.06
user       207.13
sys          9.40
          2300968960  maximum resident set size
                   0  average shared memory size
                   0  average unshared data size
                   0  average unshared stack size
              102195  page reclaims
               73193  page faults
                   0  swaps
                   0  block input operations
                   0  block output operations
                   0  messages sent
                   0  messages received
                   0  signals received
                 522  voluntary context switches
              411098  involuntary context switches
       2061342157532  instructions retired
        565766334356  cycles elapsed
          1031016928  peak memory footprint
```

And the same with git.
```
➜  gitoxide git:(main) /usr/bin/time -lp  git -c pack.threads=8 index-pack  -v --show-resolving-progress tests/fixtures/repos/linux.git/objects/pack/pack-3ee05b0f4e4c2cb59757c95c68e2d13c0a491289.pack

Indexing objects: 100% (7600359/7600359), done.
Resolving deltas: 100% (6396700/6396700), done.
3ee05b0f4e4c2cb59757c95c68e2d13c0a491289
real        33.43
user        89.95
sys         17.38
          1056047104  maximum resident set size
                   0  average shared memory size
                   0  average unshared data size
                   0  average unshared stack size
               98602  page reclaims
                   0  page faults
                   0  swaps
                   0  block input operations
                   0  block output operations
                   0  messages sent
                   0  messages received
                  30  signals received
                 472  voluntary context switches
             2054111  involuntary context switches
        496454303072  instructions retired
        295875505502  cycles elapsed
          1055084848  peak memory footprint
```

Nothing effectively changed but the new measurements contain insights into memory usage.
If gitoxide could use an equally fast Sha1 implementation it could prpbably save 100 CPU seconds.
